### DISCUSSION TITLE
Quick Start Guide?

### DISCUSSION THREAD

**Discussion by evolsoulx** [tphakala/birdnet-go#81] [General]
Alrighty, this is like, exactly what i need. but i'm more a front end web dev guy. I'm literally stuck at the very first part of this, as in i need to compile it, but i'm just clueless here. 

Tried using gpt to help a bit, but i think someone else could explain it quicker, and also, i think a lot of people could use a quick start guide?

Any way, for me specifically, i have access to a completely mine windows 11 machine, and a work limited mac. I think i can install whatever i need on the mac, and the windows machine is open. So what all do i need here?

**Comment by isZumpo** [comment]
Hi,

I agree that the documentation for getting started in terms of development could be improved. Do you have some specific steps i mind? Assuming that you want to make changes to the repository rather than just running the application? Because if you just want to run it, then I do recommend looking at the docker section of the readme. 

If you would like to build the project for development purposes. Then the easiest metod to figure out how to build locally might be to take a look at the existing documentation as well as the dockerfile where all the dependencies etc can be seen.

What I personally have been doing is to make use of the dockerfile to build local changes as well. Do this in a terminal inside the root of the repository:

```
 docker build .
```
Take the docker hash/sha that gets spit out at the end and enter it into the docker run command:
```
docker run -ti \
  -p 8080:8080 \
  --device /dev/snd \
  <sha output from docker build>
```
(see readme for more options in the docker run command such as volumes)

You should now be able to visit localhost:8080 and see your local changes. Then whenever you modify something you will have to repeat the above steps again.

I hope that any of the two steps above help you gets started. Of course, feel free to ask if there is anything else too.

**Comment by evolsoulx** [comment]
TIL what docker is 🤣.

Ok i've got the sha, is this what we say when a program is built and signed? So i just did that for myself right?

So i'm stuck on this next docker run part. This is 100% me just not understanding docker, i swear once i can get it going i can do something worthwhile haha.

where is that docker run command actually... being run? windows cmd, mac terminal, is there some kind of console i'm not immediately seeing in docker?

doh. the new line for windows command is ^, getting further, system cannot find the path, obviously now i need to update those lol.

... and now i remember powershell is a thing.

powershell command for windows to list audio devices:
Get-WmiObject Win32_SoundDevice | Select-Object Name, ProductName, DeviceID, Manufacturer

**Comment by evolsoulx** [comment]
hah yeah, at this point, i can't even get it just up and running. 'something' here doesn't make sense, but i don't know what's blocking me

**Comment by isZumpo** [comment]
So it appears that there are many things happening at once here then :)


> where is that docker run command actually... being run? windows cmd, mac terminal, is there some kind of console i'm not immediately seeing in docker?

Docker is a command line utility like any other. Meaning that you will be executing it in your terminal of choice.

First of all it might be a good idea to verify that you can run docker at all on your windows setup. Maybe follow some sort of beginners guide to get started. Windows is a bit tricky and it might suggest you to install WSL etc, leaving you wishing for that you had just installed Linux rather than Windows from the start... Probably easier to install docker on your mac if you just want to get started quickly.

Once you have verified that you can run some sort of docker image on your computer and feel ready to get started with birdnet-go. This should be the most simple docker command for that:

> docker run -ti -p 8080:8080 --privileged ghcr.io/tphakala/birdnet-go:latest

(Adding the privileged flag above since I am not sure how mounting of audio devices works in windows...)

**Reply by evolsoulx** [reply]
thank GOD it is that simple. i've been bashing my head at it for a bit hah

I do get an error and i can see in docker  (now, i see it's been adding stuff all day, oops.) a few containers, an image for ghcr.io......, some volumes that look like sha (48kb), gonna tinker for a few trying to get this to work, but here's the error incase you just see something obvious:

PS C:\Users\evols\Desktop\lsd\bnet> docker run -ti -p 8080:8080 --privileged ghcr.io/tphakala/birdnet-go:latest
Created default config file at: /root/.config/birdnet-go/config.yaml
BirdNET GLOBAL 6K V2.4 FP32 model initialized, using 16 threads of available 16 CPUs
Starting BirdNET-Go Analyzer in realtime mode
Threshold: 0.8, sensitivity: 1, interval: 15
[2024-03-25T21:25:44Z] [INFO] ⇨ http server started on [::]:8080
Capture Devices
    0: Discard all samples (playback) or generate zero samples (capture), 6e756c6c, [ok]
ALSA lib confmisc.c:855:(parse_card) cannot find card '0'
ALSA lib conf.c:5180:(_snd_config_evaluate) function snd_func_card_inum returned error: No such file or directory
ALSA lib confmisc.c:422:(snd_func_concat) error evaluating strings
ALSA lib conf.c:5180:(_snd_config_evaluate) function snd_func_concat returned error: No such file or directory
ALSA lib confmisc.c:1334:(snd_func_refer) error evaluating name
ALSA lib conf.c:5180:(_snd_config_evaluate) function snd_func_refer returned error: No such file or directory
ALSA lib conf.c:5703:(snd_config_expand) Evaluate error: No such file or directory
ALSA lib pcm.c:2666:(snd_pcm_open_noupdate) Unknown PCM default
ALSA lib confmisc.c:855:(parse_card) cannot find card '0'
ALSA lib conf.c:5180:(_snd_config_evaluate) function snd_func_card_id returned error: No such file or directory
ALSA lib confmisc.c:422:(snd_func_concat) error evaluating strings
ALSA lib conf.c:5180:(_snd_config_evaluate) function snd_func_concat returned error: No such file or directory
ALSA lib confmisc.c:1334:(snd_func_refer) error evaluating name
ALSA lib conf.c:5180:(_snd_config_evaluate) function snd_func_refer returned error: No such file or directory
ALSA lib conf.c:5703:(snd_config_expand) Evaluate error: No such file or directory
ALSA lib pcm.c:2666:(snd_pcm_open_noupdate) Unknown PCM dsnoop
ALSA lib confmisc.c:855:(parse_card) cannot find card '0'
ALSA lib conf.c:5180:(_snd_config_evaluate) function snd_func_card_id returned error: No such file or directory
ALSA lib confmisc.c:422:(snd_func_concat) error evaluating strings
ALSA lib conf.c:5180:(_snd_config_evaluate) function snd_func_concat returned error: No such file or directory
ALSA lib confmisc.c:1334:(snd_func_refer) error evaluating name
ALSA lib conf.c:5180:(_snd_config_evaluate) function snd_func_refer returned error: No such file or directory
ALSA lib conf.c:5703:(snd_config_expand) Evaluate error: No such file or directory
ALSA lib pcm.c:2666:(snd_pcm_open_noupdate) Unknown PCM dsnoop:0
ALSA lib confmisc.c:855:(parse_card) cannot find card '0'
ALSA lib conf.c:5180:(_snd_config_evaluate) function snd_func_card_id returned error: No such file or directory
ALSA lib confmisc.c:422:(snd_func_concat) error evaluating strings
ALSA lib conf.c:5180:(_snd_config_evaluate) function snd_func_concat returned error: No such file or directory
ALSA lib confmisc.c:1334:(snd_func_refer) error evaluating name
ALSA lib conf.c:5180:(_snd_config_evaluate) function snd_func_refer returned error: No such file or directory
ALSA lib conf.c:5703:(snd_config_expand) Evaluate error: No such file or directory
ALSA lib pcm.c:2666:(snd_pcm_open_noupdate) Unknown PCM dsnoop:0,0
ALSA lib confmisc.c:165:(snd_config_get_card) Cannot get card index for 0
ALSA lib confmisc.c:165:(snd_config_get_card) Cannot get card index for 0
ALSA lib confmisc.c:165:(snd_config_get_card) Cannot get card index for 0
2024/03/25 21:25:44 Device init failed miniaudio: Failed to open backend device
2024/03/25 21:25:44 ERROR: User 'root' is not member of audio group, add user to audio group by executing
2024/03/25 21:25:44 sudo usermod -a -G audio root

edit: note i did try using that sudo in powershell and after typing bash in powershell, cmd not found in powershell, with bash it does something, but then gives me the same error.

**Reply by isZumpo** [reply]
Yeah I think you are running into issues with the audio devices. Seems like privileged is not enough to solve that...

Maybe @tphakala has some experience with audio devices on windows and could give some tips?

**Reply by tphakala** [reply]
I have never used Docker on Windows, I also would recommend against using Docker images on Windows platform for now as I have no skills to help with that. I will create a new release during eastern holiday at latest and that will include a new binary build for Windows.

**Reply by tphakala** [reply]
I finally managed to get Github Workflow working for automatic build for releases, took me several days to get it going .. Anyhow there is now a fresh binary release for Windows at https://github.com/tphakala/birdnet-go/releases/tag/v0.5.0

Zip contains two files, birdnet-go.exe which is a native Windows binary, (no need to use Docker or WSL), and tensorflow lite c api library libtensorflowlite_c.so, which must be included in system PATH or in same directory birdnet-go.exe is being run at.

Run .\birdnet-go.exe once to get config.yaml file created and you can configure it with notepad etc. To enable log which you can use for OBS chatlog set

    realtime:
      log:
        enabled: true
        path: path\to\chatlog.txt (this is file which you refer to in OBS)

To make spectrograms work you need to install SoX which is available for Windows at https://sourceforge.net/directory/?q=sox-14.4.2-win64.exe

Run .\birdnet-go.exe realtime and it should start listening on default audio capture device on Windows, if it attaches on wrong audio device you need to change default recording device in Windows sound settings. Web UI is available at your Windows system IP on port 8080.

**Reply by evolsoulx** [reply]
!!! Up and running! I was worried with the whole audio config thing, but you have a simple and working way for me to hook my rstp stream in with no issues!

I can't seem to get sox to work, got it installed and path set up (i can do sox --version in powershell, but beyond that, dunno), but i don't need that, the log file is what i need, and it's absolutely perfect <3.

This whole week i've been trying to set up a new box to do some things on. It's got a wonderful bloatware Acer install of Windows 8.1, which i had never used. I'm in the process, finally, of reinstalling everything and finally have it working, but once i can find a USB drive i'm going to throw linux on it and try to get this all running there.

Incredibly stupid question; Ubuntu 23 is what i want, yeah? I've never actually set up a linux box, always sounded so complicated, but looks like it's stupid easy now. That is if i can find this damn usb stick. Going to end up just buying one when i know i have like 30 lying around somewhere 🤣 

Thank you so much for your help, like i said, once i figure out all this set up crap, i can probably give you some cool frontend stuff.

**Reply by evolsoulx** [reply]
https://www.twitch.tv/squirrelville got a super first-go display; just a python window parsing the last 5 of the text file, but oh my god, you have no idea the smile on my face right now :D 

and another dumb question: the sensitivity; is lower detecting more or less sound?

**Reply by tphakala** [reply]
Sensitivity value is a setting of BirdNET model and I can't say that I fully understand it, in technical terms it is "sigmoid sensitivity" and it is part of neural networks algorithms used in machine learning. If you are seeing too few detections you may increase it and see if helps, but  increasing its value will most likely also increase number of false detections.

About what Linux to use, yes I guess Ubuntu Linux 23.10 is a solid choice,

**Reply by evolsoulx** [reply]
got ubuntu installed! gonna play with that this week and get it set up as a 'dev' box, initially just for this, but i need to figure all this stuff out, it's holding me back.

is birdnet using the same 'stuff' as ebird in terms of identifications (new to bird call/watching world, getting terms down lol)? I'm missing a lot of cardinals that Merlin is picking up from the same audio feed, but it could easily just be a sensitivity/settings thing.
